### PR TITLE
[BATTC][NTOS:PO] Fix querying of battery state

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

Implements proper querying of the battery state in NtPowerInformation

JIRA issue: [CORE-XXXX](https://jira.reactos.org/browse/CORE-XXXX)

## Proposed changes
- [BATTC] Fix IOCTL_BATTERY_QUERY_STATUS
- [NTOS:PO] Implement SystemBatteryState in NtPowerInformation

## Testbot runs (Filled in by Devs)

- [ ] KVM x86:
- [ ] KVM x64:

### CODE REVIEW

**Comment by shubhs09** [comment]
NtPowerInformation is a function in Windows (and ReactOS tries to be compatible with Windows) that lets the system or programs ask about power-related info, like the battery status.

The issue says "Implements proper querying of the battery state in NtPowerInformation". This means the current ReactOS code doesn’t correctly answer questions about the battery status when someone asks using this function.

Specifically, two things are mentioned:

Fixing IOCTL_BATTERY_QUERY_STATUS — this is a command to the battery driver to get the current battery status (like charge level, charging or not, etc.).

Implementing SystemBatteryState in NtPowerInformation — this means making sure when the system asks about the battery, ReactOS gives the right info back.

File: sdk/include/ddk/ntpoapi.h [hunk 1]
```diff
@@ -386,7 +386,8 @@ typedef struct {
   BOOLEAN BatteryPresent;
   BOOLEAN Charging;
   BOOLEAN Discharging;
-  BOOLEAN Spare1[4];
+  BOOLEAN Spare1[3];
+  UCHAR Tag;
```
Comment (HBelusca) [inline sdk/include/ddk/ntpoapi.h:390]: This needs to be modified there too:
https://github.com/reactos/reactos/blob/719ea022ec5d85a41c3df3d79ed22fe1f93c2db8/sdk/include/xdk/winnt_old.h#L4192-L4204
(See also https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-system_battery_state )

EDIT: I see this has been done in https://github.com/reactos/reactos/pull/8271/files#diff-859a5266bac59fabdcb3de3d930f98f8249aee185c12605b75c5b98dbbda8c5eR4203 , so it's fine.

File: ntoskrnl/po/power.c [hunk 2]
```diff
@@ -824,23 +933,34 @@ NtPowerInformation(IN POWER_INFORMATION_LEVEL PowerInformationLevel,
...
+                if (!NT_SUCCESS(Status))
+                {
+                    DPRINT1("Failed to query battery state: Status: 0x%08X\n", Status);
+                    return Status;
+                }
+            }
+            else
+            {
+                /* No batteries present, set defaults */
+                RtlZeroMemory(&BatteryState, sizeof(SYSTEM_BATTERY_STATE));
```
Comment (HBelusca) [inline ntoskrnl/po/power.c]:
```suggestion
                RtlZeroMemory(&BatteryState, sizeof(BatteryState));
```

File: drivers/battery/battc/battc.c [hunk 3]
```diff
@@ -272,6 +272,7 @@ BatteryClassIoctl(PVOID ClassData,
             break;
 
         case IOCTL_BATTERY_QUERY_STATUS:
+            Irp->IoStatus.Information = 0;
```
Comment (HBelusca) [inline drivers/battery/battc/battc.c]: This is already set by default above at line 218.

File: drivers/battery/battc/battc.c [hunk 4]
```diff
@@ -321,7 +320,11 @@ BatteryClassIoctl(PVOID ClassData,
 
                 BattClass->MiniportInfo.DisableStatusNotify(BattClass->MiniportInfo.Context);
             }
-            else
+
+            Status = BattClass->MiniportInfo.QueryStatus(BattClass->MiniportInfo.Context,
```
Comment (ThFabba) [inline drivers/battery/battc/battc.c:325]: I feel like we should zero out `BattStatus` before this call, just in case the miniport doesn't.
(SystemBuffer is not zero-initialized and both BATTERY_STATUS and BATTERY_WAIT_STATUS may have padding bytes, therefore we might leak uninitialized data to user mode)

**Review by GeoB99 — DISMISSED** [review]
Composite battery state reporting is already implemented in my PR (#5719). As my work collides with yours, it truly makes no sense for you to touch the PO area since all of that code will be removed once the PR will be merged in the future anyway. 

I would rather prefer that you focus on whatever fixes you brought for the Battery Class driver.

File: ntoskrnl/po/power.c [hunk 5]
```diff
@@ -784,6 +784,115 @@ NtInitiatePowerAction(IN POWER_ACTION SystemAction,
...
+    {
+        BatteryState->EstimatedTime = 0;
+    }
+
+Exit:
+    ZwClose(CompBattHandle);
+    return Status;
+}
+    
+
```
Comment (vandandalvi) [inline ntoskrnl/po/power.c:895]: Error Handling Improvements
Currently, errors from battery device IOCTLs are logged and the function returns. Consider providing more descriptive error codes or fallback logic, especially for partial failures (e.g., if information is available but status fails).
Aggregate and report all possible error sources for better diagnostics.
Code Structure and Readability

Move duplicated cleanup code (like ZwClose) to a single exit path to avoid resource leaks if future changes add more branches.
Add more comments for complex IOCTL flows to improve maintainability.
Extensibility
Abstract battery querying logic to support multiple battery devices, not just CompositeBattery.
Allow querying additional battery properties if available.
Testing and Validation
Add unit tests/mocks for the new battery querying function.
Ensure the function handles edge cases (e.g., no battery, battery removal during query).
Performance

Consider caching results if queried frequently, or make the function asynchronous if IO is slow.
Security
Validate all input and output buffer sizes and pointers to prevent kernel memory issues.

**Comment by komal1106** [comment]
Locate Files:

drivers/battery/battc/battc.c

ntos/po/power.c or relevant NtPowerInformation handler.

Edit and Add Code:

Modify the driver to return proper status values.

Implement the missing logic for SystemBatteryState.

Rebuild ReactOS:

bash
Copy
Edit
ninja bootcd
or

bash
Copy
Edit
make
(Depending on the build environment.)

Test Changes:

Run in ReactOS Testbot or local VM.

Verify battery state is now correctly displayed under Power Options or via an app that queries NtPowerInformation.

**Comment by tkreuzer** [comment]
> Composite battery state reporting is already implemented in my PR (#5719). As my work collides with yours, it truly makes no sense for you to touch the PO area since all of that code will be removed once the PR will be merged in the future anyway.
> 
> I would rather prefer that you focus on whatever fixes you brought for the Battery Class driver.

I'm fine with you replacing this code later, but your rewrite is in PR since almost 2 years and I don't think it is a good reason to prevent any fixes from going into Po right now.

**Comment by WAZIR77** [comment]
The main issue was that EVENT_BATTERY_TAG was not signaled when the battery tag was assigned (via QueryTag), causing callers to block indefinitely if WaitTime == -1. This is now fixed by explicitly calling QueryTag and signaling the event with KeSetEvent, which resolves the issue and avoids hanging threads.

**Comment by Vishwakarmanitin** [comment]
Issue: Battery charge tooltip incorrectly shows 255% instead of the expected ~97%.

🔍 Analysis:
This is likely caused by an overflow or an unsigned integer (e.g., uint8_t) being interpreted incorrectly. A value of 255 usually represents -1 in some driver contexts (e.g., “unknown or invalid reading”).
✅ Suggested Fixes:

    Clamp battery level values to 0–100%
    Add a check before displaying battery percentage:

if (batteryLevel > 100) batteryLevel = 100;

Handle special/invalid values like 255
If the value 255 means “not available” or “error,” display a fallback message like “Unknown” instead of a percentage.

Validate driver response
Add validation when reading data from cmbatt or battc modules to ensure correct formatting and avoid misinterpretation.

Improve UI tooltip logic
Before rendering the tooltip, ensure the value passed is within the valid range and not an error code.

**Review by DarkFire01 — APPROVED** [review]

**Review by HBelusca — APPROVED** [review]

**Review by GeoB99 — APPROVED** [review]
I'm fine with this getting merged as long as you specify what wine test does this PR aim to fix. I will get back on working on my PO PR and get it shipped when I can!

File: drivers/battery/battc/battc.c (hunk 4 above)
Comment (tkreuzer) [reply drivers/battery/battc/battc.c:325]: Done

**Comment by tkreuzer** [comment]
> I'm fine with this getting merged as long as you specify what wine test does this PR aim to fix. I will get back on working on my PO PR and get it shipped when I can!

It fixes ntdll_winetest:info after update to wine-10.

File: ntoskrnl/po/power.c (hunk 2 above)
Comment (tkreuzer) [reply ntoskrnl/po/power.c]: Done

File: drivers/battery/battc/battc.c (hunk 3 above)
Comment (tkreuzer) [reply drivers/battery/battc/battc.c]: Removed
